### PR TITLE
Clickhouse support

### PR DESCRIPTION
**Pull request by sartor**
Alpha version of ClickHouse DB support (https://clickhouse.yandex/)
Works:
- connect
- list of databases
- list of tables
- select (where, sort, limit)
- sql commands

Not works:
- all DDL query builders
- sort by Count(*)
- many other things

### CODE REVIEW

File: adminer/drivers/clickhouse.inc.php [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+            }
+            unset($field[1][2]);
+
+            $alter[] = ($field[1]
+                ? ($table != "" ? ($field[0] != "" ? "CHANGE " . idf_escape($field[0]) : "ADD") : " ") . " " . implode($field[1])
+                : "DROP " . idf_escape($field[0])
+            );
+        }
+        $alter = array_merge($alter, $foreign);
+        $status = ($comment !== null ? " COMMENT=" . q($comment) : "")
```
Comment (vrana) [inline adminer/drivers/clickhouse.inc.php]: What's this? It looks like it's copy/pasted from somewhere. Finish or delete.

File: adminer/drivers/clickhouse.inc.php [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+        }
+        return ($alter || $partitioning ? queries("ALTER TABLE " . table($table) . "\n" . implode(",\n", $alter) . $partitioning) : true);
+    }
+
+    function truncate_tables($tables) {
+        return apply_queries("TRUNCATE TABLE", $tables);
+    }
+
+    function drop_views($views) {
+        return queries("DROP VIEW " . implode(", ", array_map('table', $views)));
```
Comment (vrana) [inline adminer/drivers/clickhouse.inc.php]: Does this work?

File: adminer/drivers/clickhouse.inc.php [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+    }
+
+    function drop_tables($tables) {
+        return queries("DROP TABLE " . implode(", ", array_map('table', $tables)));
```
Comment (vrana) [inline adminer/drivers/clickhouse.inc.php]: What about this?

File: adminer/drivers/clickhouse.inc.php [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+                 lang('Numbers') => array("Int8" => 3, "Int16" => 5, "Int32" => 10, "Int64" => 19, "UInt8" => 3, "UInt16" => 5, "UInt32" => 10, "UInt64" => 20, "Float32" => 7, "Float64" => 16),
+                 lang('Date and time') => array("Date" => 13, "DateTime" => 20),
+                 lang('Strings') => array("String" => 0),
+                 lang('Binary') => array("FixedString" => 0),
+             ) as $key => $val) {
+        $types += $val;
+        $structured_types[$key] = array_keys($val);
+    }
+    $unsigned = array();
+    $operators = array("=", "<", ">", "<=", ">=", "!=", "~", "!~", "LIKE", "LIKE %%", "IN", "IS NULL", "NOT LIKE", "NOT IN", "IS NOT NULL"); // no "SQL" to avoid SQL injection
```
Comment (vrana) [inline adminer/drivers/clickhouse.inc.php]: Do `SELECT` queries allow data modification in ClickHouse?

Comment (sartor) [reply adminer/drivers/clickhouse.inc.php]: I am not sure, but looks like it is not possible. Can I add "SQL" operator?

Comment (vrana) [reply adminer/drivers/clickhouse.inc.php]: Yes.

**Comment by vrana** [comment]
Pull the current version so that you don't work on the old one.

**Pull request closed without merging**